LanguageFeatures/Patterns/matching_cast_A01_t01.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion At runtime, a pattern is matched against a value. This determines
+/// whether or not the match fails and the pattern refutes the value. If the
+/// match succeeds, the pattern may also destructure data from the object or
+/// bind variables.
+///
+/// Refutable patterns usually occur in a context where match refutation causes
+/// execution to skip over the body of code where any variables bound by the
+/// pattern are in scope. If a pattern match failure occurs in an irrefutable
+/// context, a runtime error is thrown.
+///
+/// To match a pattern p against a value v:
+/// ...
+/// Cast:
+/// i. If the runtime type of v is not a subtype of the cast type of p then
+///   throw a runtime error. Note that we throw even if this appears in a
+///   matching context. The intent of this pattern is to assert that a value
+///   must have some type.
+/// ii. Otherwise, match the inner pattern against v.
+///
+/// @description Checks that a run-time error is thrown if runtime type of `v`
+/// is not subtype of the cast type of `p`. Otherwise the inner pattern is
+/// matched against `v`
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/expect.dart";
+
+String test1(Object o) {
+  switch (o) {
+    case <num>[var v] as List<int>:
+      return "match";
+    default:
+      return "no match";
+  }
+}
+
+String test2(Object o) {
+  if (o case <num>[var v] as List<int>) {
+    return "match";
+  }
+  return "no match";
+}
+
+String test3(Object o) =>
+  switch (o) {
+    <num>[var v] as List<int> => "match",
+    _ => "no match"
+  };
+
+main() {
+  Expect.throws(() {
+    var (<num>[v] as List<int>) = [3.14];
+  });
+  var (<num>[v] as List<int>) = [42];
```

Comment:
I think it would be helpful to have an explicit type argument on the list literal in the initializing expression, because this simplifies the reasoning about what we're testing: Then we know immediately that the initializing expression evaluates to a `List<int>`, and then we know that we're testing the step where a `List<int>` is checked for being a `List<int>` at run time.

It's the usual consideration about testing a specific step in a focused and precise manner (and in particular: not testing type schema computation or initializing expression inference).

Also, a reader would need to know the rules quite precisely in order to avoid some misunderstandings, e.g., "the pattern yields the type `List<num>`, so we probably infer the initializing expression as `List<num>`". This would confuse the situation because we might then expect the cast to fail.

The previous case (line 57) would actually be clarified by having a type argument on the initializing expression as well (but that case is a bit less confusing because we couldn't possibly end up with an inferred type argument on the initializing expression that would make the cast succeed).